Create security blog draft (#6470)
sourcegraph/about: content/blogposts/2023/incident-draft.md

---
title: "Security update: Incident involving unauthorized admin access"
authors:
  - name: Diego Comas, Head of Security
    url:
publishDate: 2023-08-30T10:00-07:00
description: "Sourcegraph experienced a security incident on August 30, 2023. This blog describes the details, timeline, and impact of the incident."
tags: [blog]
slug: "security-update-august-2023"
published: true
---

<br/>

**TL;DR: Sourcegraph experienced a security incident that allowed a single attacker to access some data on Sourcegraph.com. This was limited to:**

- Paid customers: 
  - The license key recipient’s name and email address. 

>   - A subset of customers’ Sourcegraph license keys may have been accessed (note that license keys _do not_ enable access to Sourcegraph instances). We are reaching out directly to those who may have been impacted to rotate license keys. 

Reviewer: ```suggestion
  - A small subset of customers’ Sourcegraph license keys may have been accessed (note that license keys _do not_ enable access to Sourcegraph instances). We are reaching out directly to those who may have been impacted to rotate license keys. 
```

- Community users: 
  - Sourcegraph account email addresses. No action is required. 

<br/>

**No other customer info, including private code, emails, passwords, usernames, or other PII, was accessible.**

<br/>

### Background

> Sourcegraph experienced a security incident on August 30, 2023 where a malicious actor used a leaked admin access token in our public Sourcegraph instance at Sourcegraph.com. The malicious external user used their privileges to increase API rate limits for users. 

Reviewer: ```suggestion
Sourcegraph experienced a security incident on August 30, 2023 where a malicious actor used a leaked admin access token in our public Sourcegraph instance at Sourcegraph.com. The malicious external user used their privileges to increase API rate limits for a small number of users. 
```

On August 30, 2023 our team noticed a significant increase in API usage and began investigating the cause. 

<img
    style={{marginTop: "20px", marginBottom: "10px", width: "60%"}}
    src="https://storage.googleapis.com/sourcegraph-assets/blog/security-update-unauthorized-admin-access/api-usage-spike.png"
    alt="A chart of Sourcegraph's API usage spike"
/>
<br/>

> The spike in usage was ruled as isolated and inorganic and our security, engineering, and support teams quickly assembled to understand what was going on.

Reviewer: Capitalise team names?

> Our security team identified a code commit from July 14 where a site-admin access token was accidentally leaked in a PR and was leveraged to impersonate a user to gain access to the administrative console of our system. 

Reviewer: pull request?

Author: ```suggestion
Our security team identified a code commit from July 14 where a site-admin access token was accidentally leaked in a pull request and was leveraged to impersonate a user to gain access to the administrative console of our system. 
```

In the spirit of transparency, we want to share the full timeline of the incident and what we have done to resolve this incident, as well as additional steps we’re taking to prevent this kind of leak in the future.

### Timeline

> On July 14, 2023 _(2023-07-14 22:01:00 UTC)_ a Sourcegraph engineer accidentally committed a code change that had an active site-admin access token. The site-admin access token had broad privileges to view and modify account information on Sourcegraph.com.

Reviewer: ```suggestion
On July 14, 2023 _(2023-07-14 22:01:00 UTC)_ a Sourcegraph engineer accidentally committed a code change that contained an active site-admin access token. The site-admin access token had broad privileges to view and modify account information on Sourcegraph.com.
```

The impact of the malicious user having admin access was limited to a subset of:

- **Paid Customers**
  - The license key recipient’s name and email address
  - Sourcegraph license key

<br/>

- **Free-Tier Community Users**
  - Email addresses

<br/>

We have no indication that any of this data was viewed, modified, or copied, but the malicious user could have viewed license key recipient’s emails and community user email addresses as they navigated the admin dashboard.

**Important Note: Customers’ private data or code was not viewed during this incident. Customer private data and code resides in isolated environments and were therefore not impacted by this event.**

### How we're mitigating

As soon as we understood the scope of the incident we took the following steps:

- Identified the malicious account and fully revoked its access
- Proactively rotated a subset of Sourcegraph customer license keys that may have been viewed
- Temporarily reduced the rate limits for all free community users
- Created new processes and tests and will continue to monitor for malicious activity and abuse

> Expanding our secret scanning through additional static analysis tests will ensure we can better detect and prevent this kind of leak in the future.

Reviewer: This feels a bit weird straight after the bullets - make it clear we will do this.
